core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.cost.TableStatsProvider;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.metadata.Metadata;
+import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.type.ArrayType;
+import io.trino.spi.type.IntegerType;
+import io.trino.sql.planner.FunctionCallBuilder;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanVisitor;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.SortNode;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.tree.ArrayConstructor;
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.DataType;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.IfExpression;
+import io.trino.sql.tree.InListExpression;
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LongLiteral;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.StringLiteral;
+import io.trino.sql.tree.SymbolReference;
+import oshi.util.tuples.Pair;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.SystemSessionProperties.getSkewedJoinMetadata;
+import static io.trino.SystemSessionProperties.getSkewedJoinReplicationFactor;
+import static io.trino.sql.analyzer.TypeSignatureTranslator.toSqlType;
+import static java.util.Objects.requireNonNull;
+import static io.trino.sql.planner.plan.ChildReplacer.replaceChildren;
+
+class SkewedContext
+{
+    public final DataType keyType;
+    public final List<String> values;
+
+    public SkewedContext(List<String> values, DataType keyType)
+    {
+        this.keyType = keyType;
+        this.values = values;
+    }
+}
+
+public class SkewJoinOptimizer
```

Comment:
Could you make it an iterative optimizer rule?

Reply:
I don't think i can make this a rule yet, i need to be able to traverse the graph and collect symbols that might be skewed.

Reply:
> I don't think i can make this a rule yet, i need to be able to traverse the graph and collect symbols that might be skewed.

You could use `StatsCalculator` from a rule to identify how many distinct values given join symbols has

Reply:
didn't know about that. I'm not sure how it would work though, i need to know that the symbol is the same as the column the user has specified.

Reply:
> i need to know that the symbol is the same as the column the user has specified.

You don't need user input.

You can use `io.trino.cost.StatsCalculator#calculateStats` to get statistics of join source nodes. Then you can check `io.trino.cost.SymbolStatsEstimate#getDistinctValuesCount` for join key symbol.
If `SymbolStatsEstimate#getDistinctValuesCount` is lower than some threshold (e.g. `number of nodes * task_concurrency`) then you could trigger your rule.

Reply:
> didn't know about that. I'm not sure how it would work though, i need to know that the symbol is the same as the column the user has specified.

You are right. You need actual skewed values which CBO cannot provide. @martint are we fine adding session property like `SKEWED_JOIN_METADATA` of we would like to have something more adaptive?